tests/phpunit/tests/option/option.php

```diff
@@ -368,4 +368,131 @@ public function data_option_autoloading() {
+
+		$this->assertFalse( $updated, 'update_option should not return true when values are loosely equal.' );
+		$this->assertSame( $num_queries, get_num_queries(), 'The number of database queries should not change.' );
+	}
+
+	/**
+	 * @ticket 22192
+	 *
+	 * @covers ::update_option
+	 *
+	 * @dataProvider data_update_option_type_juggling
+	 *
+	 * @param mixed $old_value One of the values to compare.
+	 * @param mixed $new_value The other value to compare.
+	 */
+	public function test_update_option_should_hit_cache_when_loosely_equal_to_existing_value_and_cached_values_are_pulled_from_the_database( $old_value, $new_value ) {
+		add_option( 'foo', $old_value );
+		wp_cache_delete( 'alloptions', 'options' );
+		wp_load_alloptions();
+
+		$num_queries = get_num_queries();
+
+		$updated = update_option( 'foo', $new_value );
+
+		$this->assertSame( (string) $old_value, get_option( 'foo' ), 'The value should not have been changed.' );
+		$this->assertFalse( $updated, 'update_option should not return true when values are loosely equal.' );
+		$this->assertSame( $num_queries, get_num_queries(), 'The number of database queries should not change.' );
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array
+	 */
+	public function data_update_option_type_juggling() {
+		return array(
+			// Truthy.
+			array( '1', '1' ),
+			array( '1', 1 ),
+			array( '1', 1.0 ),
+			array( '1', true ),
+			array( 1, '1' ),
+			array( 1, 1 ),
+			array( 1, 1.0 ),
+			array( 1, true ),
+			array( 1.0, '1' ),
+			array( 1.0, 1 ),
+			array( 1.0, 1.0 ),
+			array( 1.0, true ),
+			array( true, '1' ),
+			array( true, 1 ),
+			array( true, 1 ),
+			array( true, true ),
+
+			// Falsey.
+			array( '0', '0' ),
+			array( '0', 0 ),
+			array( '0', 0.0 ),
+			array( '0', false ),
+			array( 0, '0' ),
+			array( 0, 0 ),
+			array( 0, 0.0 ),
+			array( 0, false ),
+			array( 0.0, '0' ),
+			array( 0.0, 0 ),
+			array( 0.0, 0.0 ),
+			array( 0.0, false ),
+		);
+	}
+
+	/**
+	 * @ticket 22192
+	 *
+	 * @covers ::update_option
+	 *
+	 * @dataProvider data_update_option_values_loosely_equal_to_false
+	 *
+	 * @param mixed $old_value One of the values to compare.
+	 * @param mixed $new_value The other value to compare.
+	 * @param bool  $expected  The expected result.
+	 */
+	public function test_update_option_with_falsey_values( $old_value, $new_value, $expected ) {
+		add_option( 'foo', $old_value );
+		$num_queries = get_num_queries();
+
+		$updated = update_option( 'foo', $new_value );
+
+		if ( $expected ) {
+			$this->assertSame( $new_value, get_option( 'foo' ), 'The value should have been changed.' );
+		} else {
+			$this->assertSame( $old_value, get_option( 'foo' ), 'The value should not have been changed.' );
+		}
+		$this->assertSame( $expected, $updated, 'update_option should have returned ' . ( $expected ? 'true' : 'false' ) );
+		$this->assertSame( 1, get_num_queries() - $num_queries, 'There should be one additional database query to update the option.' );
+	}
+
```

Comment:
@felixarntz In side PR we have `add_option( 'foo', false )` related test that we removed in https://github.com/WordPress/wordpress-develop/pull/5139/commits/bf1ef31f139f1d4c35abb0896d0d02cd5800c1da. Please correct me if i'm wrong or miss anything here.

Reply:
@mukeshpanchal27 I mean these two tests:
* https://github.com/WordPress/wordpress-develop/commit/bf1ef31f139f1d4c35abb0896d0d02cd5800c1da#diff-2f4fd4e950f239051cbb449ee9cc745f348e2e08e1534aec134adaddf7533a79L527
* https://github.com/WordPress/wordpress-develop/commit/bf1ef31f139f1d4c35abb0896d0d02cd5800c1da#diff-2f4fd4e950f239051cbb449ee9cc745f348e2e08e1534aec134adaddf7533a79L551

These tests were indeed incorrect in the way that they were written because they were asserting that a non-existent option is _not_ added - but it _should_ be added.

So rather than not having those tests, we should "fix" them by testing the opposite: If you call `update_option()` passing the same value as the default, it _should_ be set in the database. We need two tests for that, one to cover the default default of `false`, and another to cover the default using the filter.